crates/uv-resolver/src/resolver/environment.rs

```diff
@@ -0,0 +1,420 @@
+    /// `Requires-Python`) is `>=3.10`, then this will "narrow" the requirement
+    /// to `>=3.12`, corresponding to the marker expression describing this
+    /// fork.
+    ///
+    /// If this environment is not a fork, then this returns `None`.
+    pub(crate) fn narrow_python_requirement(
+        &self,
+        python_requirement: &PythonRequirement,
+    ) -> Option<PythonRequirement> {
+        Some(python_requirement.narrow(&self.requires_python_range()?)?)
+    }
+
+    /// Returns a message formatted for end users representing a fork in the
+    /// resolver.
+    ///
+    /// If this resolver environment does not correspond to a particular fork,
+    /// then `None` is returned.
+    ///
+    /// This is useful in contexts where one wants to display a message
+    /// relating to a particular fork, but either no message or an entirely
+    /// different message when this isn't a fork.
+    pub(crate) fn end_user_fork_display(&self) -> Option<impl std::fmt::Display + '_> {
+        match self.kind {
+            Kind::Specific { .. } => None,
+            Kind::Universal { ref markers, .. } => {
+                if markers.is_true() {
+                    None
+                } else {
+                    Some(format!("split ({markers:?})"))
+                }
+            }
+        }
+    }
+
+    /// Returns the marker expression corresponding to the fork that is
+    /// represented by this resolver environment.
+    ///
+    /// This returns `None` when this does not correspond to a fork.
+    pub(crate) fn try_markers(&self) -> Option<&MarkerTree> {
+        match self.kind {
+            Kind::Specific { .. } => None,
+            Kind::Universal { ref markers, .. } => {
+                if markers.is_true() {
+                    None
+                } else {
+                    Some(markers)
+                }
+            }
+        }
+    }
+
+    /// Returns a requires-python version range derived from the marker
+    /// expression describing this resolver environment.
+    ///
+    /// When this isn't a fork, then there is nothing to constrain and thus
+    /// `None` is returned.
+    fn requires_python_range(&self) -> Option<RequiresPythonRange> {
+        crate::marker::requires_python(self.try_markers()?)
+    }
+}
+
+/// A user visible representation of a resolver environment.
+///
+/// This is most useful in error and log messages.
+impl std::fmt::Display for ResolverEnvironment {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self.kind {
+            Kind::Specific { .. } => write!(f, "Marker Environment"),
+            Kind::Universal { ref markers, .. } => {
+                if markers.is_true() {
+                    write!(f, "All Marker Environments")
+                } else {
+                    write!(f, "Split `{markers:?}`")
+                }
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
Are we doing unit tests inline or in an extra file?

Reply:
I think this is idiomatic for unit tests. And we do it pretty frequently?

(Feel free to start a thread on Discord about this. I'm curious about your thoughts on this.)

Reply:
i prefer the inline tests, but all other tests were moved out in the PR that merged the integration tests, I'm asking for consistency's sake

Reply:
Yeah I think this is no longer idiomatic for us for performance reasons? I don't know all the reasoning.

Reply:
Hmmm. Yeah I get the thinking with merging the integration tests, but I don't yet grok the thinking around unit tests being inline or in a separate file.